kernel/typeops.ml

```diff
@@ -649,113 +649,415 @@ let check_let_annot env s x c t =
-      let c', ct = execute env c in
+      let c', ct = execute tbl env c in
       let r', ty = type_of_projection env p c' ct in
       assert (Sorts.relevance_equal r r');
-      let cstr = if c == c' then cstr else mkProj (p,r,c') in
+      let cstr = if self c == c' then self cstr else mkProj (p,r,c') in
       cstr, ty
 
     (* Lambda calculus operators *)
     | App (f,args) ->
-      let args', argst = execute_array env args in
+      let args', argst = execute_array tbl env args in
         let f', ft =
-          match kind f with
+          match HConstr.kind f with
           | Ind ind when Environ.template_polymorphic_pind ind env ->
-            f, type_of_inductive_knowing_parameters env ind args' argst
+            self f, type_of_inductive_knowing_parameters env ind args' argst
           | _ ->
             (* No template polymorphism *)
-            execute env f
+            execute tbl env f
         in
-        let cstr = if f == f' && args == args' then cstr else mkApp (f',args') in
+        let cstr = if self f == f' && is_self_array args args'
+          then self cstr else mkApp (f',args') in
         cstr, type_of_apply env f' ft args' argst
 
     | Lambda (name,c1,c2) ->
-      let c1', s = execute_is_type env c1 in
+      let c1', s = execute_is_type tbl env c1 in
       let name' = check_assum_annot env s name c1' in
       let env1 = push_rel (LocalAssum (name',c1')) env in
-      let c2', c2t = execute env1 c2 in
-      let cstr = if name == name' && c1 == c1' && c2 == c2' then cstr else mkLambda(name',c1',c2') in
-      cstr, type_of_abstraction env name' c1 c2t
+      let c2', c2t = execute tbl env1 c2 in
+      let cstr = if name == name' && self c1 == c1' && self c2 == c2'
+        then self cstr else mkLambda(name',c1',c2') in
+      cstr, type_of_abstraction env name' c1' c2t
 
     | Prod (name,c1,c2) ->
-      let c1', vars = execute_is_type env c1 in
+      let c1', vars = execute_is_type tbl env c1 in
       let name' = check_assum_annot env vars name c1' in
       let env1 = push_rel (LocalAssum (name',c1')) env in
-      let c2', vars' = execute_is_type env1 c2 in
-      let cstr = if name == name' && c1 == c1' && c2 == c2' then cstr else mkProd(name',c1',c2') in
+      let c2', vars' = execute_is_type tbl env1 c2 in
+      let cstr = if name == name' && self c1 == c1' && self c2 == c2'
+        then self cstr else mkProd(name',c1',c2') in
       cstr, type_of_product env name' vars vars'
 
     | LetIn (name,c1,c2,c3) ->
-      let c1', c1t = execute env c1 in
-      let c2', c2s = execute_is_type env c2 in
+      let c1', c1t = execute tbl env c1 in
+      let c2', c2s = execute_is_type tbl env c2 in
       let name' = check_let_annot env c2s name c1' c2' in
       let () = check_cast env c1' c1t DEFAULTcast c2' in
       let env1 = push_rel (LocalDef (name',c1',c2')) env in
-      let c3', c3t = execute env1 c3 in
-      let cstr = if name == name' && c1 == c1' && c2 == c2' && c3 == c3' then cstr
```

Comment:
dealing with this code got pretty annoying, maybe time to make bad relevance in kernel a hard error then it can go away